core/peerset/src/lib.rs

```diff
@@ -306,11 +348,90 @@ impl PeersetMut {
 	}
 }
 
-impl Stream for PeersetMut {
+impl Stream for Peerset {
 	type Item = Message;
 	type Error = ();
 
 	fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
-		self.rx.poll()
+		loop {
```

Comment:
Do I understand it correctly that the `loop` is necessary because the various calls to `self.message_queue.push_back` will not wake-up the current task? 

Have you considered instead having the various `on_` methods return a `Option<Self::Item>` directly? Since `if let Some(message)` immediately returns, I assume there isn't a case where any of the `on_` methods will create more than one message?

Reply:
> Do I understand it correctly that the `loop` is necessary because the various calls to `self.message_queue.push_back` will not wake-up the current task?

It's necessary, cause `on_` methods may add multiple messages or may not add them at all. In the later case, we need to process next `Action`.

> Have you considered instead having the various `on_` methods return a `Option<Self::Item>` directly? Since `if let Some(message)` immediately returns, I assume there isn't a case where any of the `on_` methods will create more than one message?

Yes, I considered it, but it's less efficient and may be confusing. 

- `on_` methods would require `&mut self` even if they don't touch `self.message_queue`. So we would have an `on_` method that mutates `self` and returns the result that should be always appended to `self.message_queue`. It would be confusing and could lead to incorrect behaviours
- `on_` methods would have to return `Vec<Self::Item>` so we would have an additional allocation of on a heap of something that is rewritten to `self.message_queue` a second after

Reply:
Ok, I see, wouldn't it be possible for messsages in `message_queue` to be stuck there in case nothing new is sent on the receiver? Let's say one of the `_on` methods adds 2 items to `message_queue`, the first one get's returned by the `if let Some(message)` and the second one is still inside `message_queue`. If nothing wakes up the current task, the second message in `message_qeue` could never be handled?

Reply:
> Ok, I see, wouldn't it be possible for messsages in `message_queue` to be stuck there in case nothing new is sent on the receiver?

It's a good question, but no.

> Let's say one of the `_on` methods adds 2 items to `message_queue`, the first one get's returned by the `if let Some(message)` and the second one is still inside `message_queue`. If nothing wakes up the current task, the second message in `message_qeue` could never be handled? Maybe schedule a wake-up if the `message_queue` is non-empty before returning from `poll`

If a stream returns a single item, it will be waken up subsequently until it returns `Async::NotReady` or `Async::Ready(None)`. So if we push multiple messages to `message_queue`, it will be polled until all the items from `message_queue` are drained

Reply:
Ok, I didn't know one could rely on the stream being woken-up later by returning `Async::Ready(Some)`. I thought one had to have another way to register the current task for being woken-up, for example in the current case by calling `self.rx.poll()`, which registers the current task for being woken-up if a message is sent on the corrseponding sender.